libnmstate/metadata.py

```diff
@@ -181,3 +202,127 @@ def _generate_route_metadata(desired_state):
+                                   (ipv4_iface, Interface.IPV4, ipv4_server)):
+        if servers:
+            if not iface:
+                raise NmstateValueError(
+                    'Failed to find suitable interface for saving DNS '
+                    'name servers: %s' % servers)
+            if iface not in desired_state.interfaces:
+                desired_state.include_changed_interfaces(current_state, [iface])
+            dns_meta = {
+                DNS_PRIORITY: DNS_DEFAULT_PRIORITY_FOR_STATIC,
+                DNS.SERVER: servers,
+                DNS.SEARCH: []
+            }
+            if need_dns_priority:
+                if family == Interface.IPV6:
+                    dns_meta[DNS_PRIORITY] = DNS_ORDERING_IPV6_PRIORITY
+                else:
+                    dns_meta[DNS_PRIORITY] = DNS_ORDERING_IPV4_PRIORITY
+            if not search_saved:
+                dns_meta[DNS.SEARCH] = searches
+                search_saved = True
+            iface_state = desired_state.interfaces[iface]
+            if family not in iface_state:
+                iface_state[family] = {DNS_METADATA: dns_meta}
+            else:
+                iface_state[family][DNS_METADATA] = dns_meta
+    # Some interfaces might old DNS configuration which is not included
+    # in desired_state, we need to add them into desired_state in order to
+    # clear its DNS configurations
+    old_dns_ifaces = nm_dns.get_dns_config_iface_names()
+    desired_state.include_changed_interfaces(current_state, old_dns_ifaces)
+
+
+def _choose_iface_for_dns(desired_state, current_state):
+    """
+    Find out the interface to store the DNS configuration:
+        * Static gateway configured
+        * DHCP with auto-dns false
+    Return two iface_names, first is for ipv4, second for ipv6.
+    """
+    ipv6_iface = None
+    ipv4_iface = None
+    new_state = copy.deepcopy(desired_state)
+    new_state.merge_interfaces(current_state)
+    for iface_name, routes in six.viewitems(new_state.config_iface_routes):
+        for route in routes:
+            if ipv6_iface and ipv4_iface:
+                return ipv4_iface, ipv6_iface
+            if not ipv6_iface and \
+               route[Route.DESTINATION] == IPV6_DEFAULT_GATEWAY_DESTINATION:
+                ipv6_iface = iface_name
+                continue
+            if not ipv4_iface and \
+               route[Route.DESTINATION] == IPV4_DEFAULT_GATEWAY_DESTINATION:
+                ipv4_iface = iface_name
+                continue
+    if not ipv4_iface:
+        ipv4_iface = _choose_auto_iface_without_auto_dns(
+            Interface.IPV4, desired_state, current_state)
+    if not ipv6_iface:
+        ipv6_iface = _choose_auto_iface_without_auto_dns(
+            Interface.IPV6, desired_state, current_state)
+    return ipv4_iface, ipv6_iface
+
+
+def _choose_auto_iface_without_auto_dns(family, desired_state, current_state):
```

Comment:
I suggest you do this also using a lookup of interfaces that met the predicate.
However, it looks much more complicated, because you need to take into account if it is not absent or IP stack disabled and many other points.
I suggest we postpone this one for later and discuss the algorithm further.